### ISSUE TITLE
How to properly make my SWA logout from a custom identity provider?

### ISSUE THREAD

**Issue by brenoalvs** [Azure/static-web-apps#594]
I saw this similar issue and I think this is a bug.
https://github.com/Azure/static-web-apps/issues/529

But I would like to understand better if this is something that will be fixed or do we really need to make the "reverse flow" to make it work. (Logout from IdP redirecting to app logout url).

**Describe the bug**
The app does not completely logs out when we visit the /logout url with a custom identity provider configured.

**To Reproduce**
1. Setup a custom identity provider
2. Visit the logout url 
3. The SWA app will be logged out but not the custom provider.
4. If you visit any url that requires authentication the app stills logged in without requesting for credentials again.
 
**Expected behavior**
Logout from the SWA and from the custom identity provider.

**Comment by mkarmark** [comment]
Hi, we are taking a look at this issue and will post about a potential mitigation plan when we figure out the issue.

**Comment by ananratt1** [comment]
I have the same issue. I can confirm that /logout doesn't really log the user out from AAD or Google.

Steps to repro:
1. Navigate to /.auth/login/aad and login using a Microsoft Account. Say NO to 'Stay signed in?' prompt.
2. Navigate to /.auth/me to see the basic information on the logged in account to prove that I'm in the logged in state.
3. Navigate to /.auth/logout to logout. Immediately, navigate to /.auth/me again to confirm that my static Web App regards me as 'logged out.' I'd see this:
{"clientPrincipal": null}
4. Navigate to /.auth/login/aad again. Microsoft login page flashes by, and I am logged right back in with the previously logged out user.

Using .auth/login/google suffers from the same issues.

**Things that sort of worked**
Any of the two actions below alone seems to make the browser forget my logged-in state:

1. Close the whole browser and relaunch it. I'd get asked to pick the previous user (and then enter the password) or choose a new user.

2. After having logged in to SWA using AAD, open a new tab in the same browser, and navigate to hotmail.com. That tab will enjoy my logged-in state from the Static Web Apps tab. I'd see my mails right away. Then I log out from the hotmail.com tab, switch back to ***.azurestaticapps.net tab and see that I am still logged into my Static Web App. Good! Then if I log out from my Static Web App and try to log back in, it has forgotten my logged-in state this time. In other words, logging out from the hotmail.com tab is somehow more powerful.

**Comment by TellonUK** [comment]
Are there plans to make this work with AzureB2C when used as a custom provider? 

A solution for AzureB2C (And most OIDC Providers) is to have .auth/logout redirect to the end session URL provided in the known good configuration, you can then pass a post_logout_redirect to get back to the .auth/login/AzureB2C page. 

If you could provide a logout URL somewhere that .auth/logout should redirect to as part of the app configuration, or in the json configuration or even pick it up from the OpenID end session property from the known good configuration then it could be configured per static web app, allowing the same code to be used for multiple instances of the application utilizing different AzureB2C providers (multiple clients)

**Comment by bzbetty** [comment]
I think b2c supplies the logout URL in its config endpoint, so shouldn't need the URL as a setting, just whether it should log you out.

**Comment by StefanCardnell** [comment]
@mkarmark 

I was banging my head against the wall trying to understand why logout was not working for my custom Azure B2C provider. 

I just tested Github, twitter and Azure B2C. Logout does not work for any - logging in again never asks for credentials. If identity provider logout only works for a small handful of providers, this really needs to be said in the docs. It is a security concern, since obviously logout does not truly log you out.

https://docs.microsoft.com/en-us/azure/active-directory-b2c/configure-authentication-in-azure-static-app

This is an entire guide for setting up SWA with Azure B2C, it doesn't mention logout won't work for most ID providers.

https://docs.microsoft.com/en-us/azure/active-directory-b2c/configure-authentication-in-azure-static-app

Same here.

**Comment by fralik** [comment]
Any news on that? I am experience the same issue on a computer located in a public space. After I log in and then log out from the app, anyone could re-log on my behalf.

**Comment by TellonUK** [comment]
My custom identity provider (Azure B2C) has a logout URL. I was able to goto that logout URL (logging out of my Azure B2C tenant) then have that redirect to the Azure SWA url for logout, which would trigger a local logout of the SWA. This allowed me to excite some cleanup code, logout of the IDP and SWA.

**Comment by MudassarAli** [comment]
Any progress here. We are experiencing the same issue.

**Comment by scafaria** [comment]
I have this issue too.

**Comment by bzbetty** [comment]
like @TellonUK we ended up redirecting the user to "https://$(tenantName).b2clogin.com/$(tenantName).onmicrosoft.com/b2c_1_login/oauth2/v2.0/logout?post_logout_redirect_uri=https%3A%2F%2F$(redirectUri)%2F.auth%2Flogout" which then redirects the user to the swa logout page. seems to work for us.

**Comment by scafaria** [comment]
Thanks for the reply @bzbetty. I'm using the built-in identity providers for Static Web Apps, e.g. "https://myurl.com/.auth/login/google". I'm not sure if that's why I don't have any luck, but when I try the URL you provided (using my tenantName and redirectUri) I receive "The resource you are looking for has been removed, had its name changed, or is temporarily unavailable."

**Comment by bzbetty** [comment]
also the b2c_1_login section of the link is the name of your logout user flow

**Comment by scafaria** [comment]
Right. I'm using the built-in auth of a static web app. So all I have is (twitter example): "https://mysite.com/.auth/login/twitter" and "https://mysite.com/.auth/logout". (also should note that "mytenantName" differs from mysite.com")

**Comment by dzidzedev** [comment]
Hi,
Is there here any fix? I have tried with WSO2 custom identity provider. I have the same issue.

**Comment by GNANASEKHARKURA** [comment]
Hi there,

It's a long pending issue, any update on this?

**Comment by phranger** [comment]
+1 Still an issue with OIDC not being able to log our or even pass parameters to logout url

**Comment by kentmor** [comment]
Is this still not resolved? I have an SWA using Entra ID authentication as outline in the documentation (https://learn.microsoft.com/en-us/azure/static-web-apps/authentication-custom?tabs=aad%2Cinvitations). 

Using the /.auth/logout route tells me I'm logged out but if I go back to my webpage I am in fact not logged out

**Comment by bospoort** [comment]
Same issue. None of the samples provided show logout working. I use the standard EntraID as Identity Provider. Log out seems to log me out and the StaticWebAppsAuthCookie is gone, but going to a route that requires authentication automatically logs me back in. Help?

**Comment by bzbetty** [comment]
@bospoort you need to manually redirect to your identity providers logout page and have it redirect back to the swa logout page.

**Comment by bospoort** [comment]
@bzbetty. That is what I do. I follow the steps [here](https://learn.microsoft.com/en-us/azure/static-web-apps/authentication-authorization). On landing on my SWA logout page, I see the cookie getting deleted. But then if I go back to a page that requires auth, the browser logs me back in.

**Comment by bzbetty** [comment]
@bospoort yeah unfortunately you can't link directly to /.auth/logout.

need to link to something more like:

https://login.microsoftonline.com/common/oauth2/logout?post_logout_redirect_uri=https%3a%2f%2fyourapp.com%2f.auth/logout

**Comment by mcollier** [comment]
Anyone find a solution? I'm encountering the same issue with my Next.js (using SSR) app, Auth0, and Azure Static Web Apps.

**Comment by byBobo** [comment]
Same for me. 
I was just testing Azure Static WebApp, so mine is a near hello world case; 
- index.html unprotected custom login page (redirecting to `/.auth/login/aad?post_login_redirect_uri=/home.html`), 
- home.html protected page
-  403.html to be redirect in case of user authenticated but not authorized. 

After logout, I can easy login again with a not authorized account and reach the "protected" home.html page.

(follow)

**Comment by jmair17** [comment]
Encountering the same problem.. Still got no working logout, is there any update on this?

**Comment by vitoras** [comment]
@mkarmark apparently there's no value in such a basic feature.

**Comment by bakicaglar** [comment]
it astonishes me that such an important "logout" functionality issue still exists and not working property on SWA.
i have used the following workaround to force calling "/.auth/logout" on logout:

```
try {
  // 1. Force the SWA backend to clear auth markers in sequence
  // Use no-cors to avoid potential redirect issues during the fetch
  await fetch('/.auth/logout', { mode: 'no-cors' })
  await fetch('/.auth/logout/complete', { mode: 'no-cors' })
} catch (e) {
  console.error('Logout fetch failed, proceeding with redirect', e)
}

// 2. Final redirect to a PUBLIC page
const logoutUrl = new URL('/.auth/logout', window.location.origin)
logoutUrl.searchParams.set('post_logout_redirect_uri', window.location.origin + '/')

window.location.href = logoutUrl.toString()
```
